flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/operators/rank/UpdatableTopNFunction.java

```diff
@@ -312,6 +357,37 @@ private void processElementWithRowNumber(RowData inputRow, Collector<RowData> ou
                 "Failed to find the sortKey, rowkey in the buffer. This should never happen");
     }
 
+    private void emitRecordsWithRowNumberIgnoreStateError(
+            RowData newRow, int newRank, RankRow oldRow, int oldRank, Collector<RowData> out) {
+        Iterator<Map.Entry<RowData, Collection<RowData>>> iterator = buffer.entrySet().iterator();
+        int currentRank = 0;
+        RowData currentRow = null;
+        RowData prevRow = null;
+
+        while (iterator.hasNext() && currentRank <= newRank) {
+            Map.Entry<RowData, Collection<RowData>> entry = iterator.next();
+            Collection<RowData> rowKeys = entry.getValue();
+            Iterator<RowData> rowKeyIter = rowKeys.iterator();
+            while (rowKeyIter.hasNext()) {
+                RowData rowKey = rowKeyIter.next();
+                currentRank += 1;
+                currentRow = rowKeyMap.get(rowKey).row;
+                if (oldRank <= currentRank) {
+                    if (currentRank == oldRank) {
+                        collectUpdateBefore(out, oldRow.row, oldRank);
+                    } else {
+                        collectUpdateBefore(out, prevRow, currentRank);
+                        collectUpdateAfter(out, prevRow, currentRank - 1);
+                        if (currentRank == newRank) {
+                            collectUpdateAfter(out, newRow, currentRank);
+                        }
+                    }
+                }
```

Comment:
What if `newSortKey > oldSortKey` but `oldRank == currentRank`? The update after message will be lost. Please also add a test about this after fixing it.

Reply:
It seems that a more proper solution to this is not to send any message if `oldRank == currentRank`.

Reply:
Good catch!  will add case to cover it.